sympy/parsing/sympy_parser.py

```diff
@@ -1074,11 +1074,19 @@ def parse_expr(s: str, local_dict: Optional[DICT] = None,
 
     code = stringify_expr(s, local_dict, global_dict, _transformations)
 
+    # Make sure we apply NFKC normalization to the input string
+    # as well as keys of local_dict to make sure Python's built
+    # in NFKC normalization, done when calling eval on code doesn't
+    # create a discrepancy between the code and local_dict keys.
+    code = unicodedata.normalize('NFKC', code)
+    local_dict_tmp = {unicodedata.normalize('NFKC', k): v
+                      for k, v in local_dict.items()}
+
     if not evaluate:
         code = compile(evaluateFalse(code), '<string>', 'eval') # type: ignore
 
     try:
-        rv = eval_expr(code, local_dict, global_dict)
+        rv = eval_expr(code, local_dict_tmp, global_dict)
         # restore neutral definitions for names
         for i in local_dict.pop(null, ()):
             local_dict[i] = null
```

Comment:
The code below modifies `local_dict` and I am not sure why.

Does it matter that `local_dict_tmp` is now being passed instead of `local_dict`?

Reply:
@oscarbenjamin you are right, there might be an important reason why the local_dict passed in is modified. I chose to create a separate, temporary dict where the characters are normalized but this might not be the right solution.

Reply:
I am starting to understand better what is happening with `local_dict` and `null` here.

One of the transformations applied prior to parsing is `auto_symbol`, which is meant to insert calls to `Symbol` or `Function` for undefined variables. In this line: https://github.com/sympy/sympy/blob/master/sympy/parsing/sympy_parser.py#L558, it collects certain names appearing in the tokens list derived from the input string and adds them to a set under the `null` key (`null`'s value is declared to be an empty string and is not allowed as a key in `local_dict` in the user input, making sure that it can only be used by the parsing logic internally).

After the expression has been evaluated, there is a piece of code to remove the null key and add any names that were inserted into its value set as keys in `local_dict` here: https://github.com/sympy/sympy/blob/master/sympy/parsing/sympy_parser.py#L1083-L1089 (one in the case of no exception and one in the case of an excetipon).

Based on how I implemented `local_dict_tmp`: it is created _after_ the `auto_symbol` transformation to `local_dict`, but code is then again using the `local_dict` when the transformation is reverted. So there is overall no effect on the original `local_dict`.

To test this, I ran all the tests and identified where the code in https://github.com/sympy/sympy/blob/master/sympy/parsing/sympy_parser.py#L1083-L1089 is triggered. This revealed `test_issue_6046` as the only test where this code is invoked: https://github.com/sympy/sympy/blob/master/sympy/core/tests/test_sympify.py#L545-L557. Based on the test, the overall behavior surrounding `local_dict` isn't changed by this PR.

Hope that's useful for reviewing the code!

Reply:
If `local_dict` is not being modified then there is no need to restore it here after calling `eval_expr`.

Reply:
The code to restore `local_dict` is generally necessary to revert changes made by `auto_symbol`. But it is not affected by / does not affect the changes in this PR.